contracts/evm/contracts/Oracle.sol

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+import "./interfaces/IOracle.sol";
+import "./interfaces/IPublisherRegistry.sol";
+import "./CurrencyManager.sol";
+import "./EntryUtils.sol";
+
+contract Oracle is CurrencyManager, EntryUtils, IOracle {
+    IPublisherRegistry public publisherRegistry;
+
+    mapping(bytes32 => bytes32[]) public oracleSourcesStorage;
+    mapping(bytes32 => mapping(bytes32 => SpotEntry)) public spotEntryStorage;
+    mapping(bytes32 => Checkpoint[]) public checkpoints;
+    mapping(bytes32 => uint256) public checkpointIndex;
+    uint256 sourcesThreshold = 1;
+
+    uint256 constant BACKWARD_TIMESTAMP_BUFFER = 3600;
+    uint256 constant FORWARD_TIMESTAMP_BUFFER = 900;
```

Comment:
We may want to tune these separately for different networks, so should probably move to a config file medium-term. For now I'd guess these are quite conservative so ok to keep